src/python/pants/engine/internals/selectors.py

```diff
@@ -72,58 +70,57 @@ def parse_product_and_subject_types(
-                    "SubjectType(constructor args), the second argument should be a constructor "
+                    "Because you are using the shorthand form Get(OutputType, "
+                    "InputType(constructor args), the second argument should be a constructor "
                     "call, like `MergeDigest(...)` or `Process(...)`."
                 )
-            if not hasattr(subject_constructor.func, "id"):
+            if not hasattr(input_constructor.func, "id"):
                 raise parse_error(
-                    "Because you are using the shorthand form Get(ProductType, "
-                    "SubjectType(constructor args), the second argument should be a top-level "
+                    "Because you are using the shorthand form Get(OutputType, "
+                    "InputType(constructor args), the second argument should be a top-level "
                     "constructor function call, like `MergeDigest(...)` or `Process(...)`, rather "
                     "than a method call."
                 )
-            return product_type, subject_constructor.func.id  # type: ignore[attr-defined]
+            return output_type, input_constructor.func.id  # type: ignore[attr-defined]
 
-        subject_type, _ = subject_args
-        if not isinstance(subject_type, ast.Name):
+        input_type, _ = input_args
+        if not isinstance(input_type, ast.Name):
             raise parse_error(
-                "Because you are using the longhand form Get(ProductType, SubjectType, "
-                "subject_instance), the second argument should be a type, like `MergeDigests` or "
+                "Because you are using the longhand form Get(OutputType, InputType, "
+                "input), the second argument should be a type, like `MergeDigests` or "
                 "`Process`."
             )
-        return product_type, subject_type.id
+        return output_type, input_type.id
 
 
 @frozen_after_init
 @dataclass(unsafe_hash=True)
-class Get(GetConstraints, Generic[_ProductType, _SubjectType]):
+class Get(GetConstraints, Generic[_Output, _Input]):
     """Asynchronous generator API.
 
     A Get can be constructed in 2 ways with two variants each:
 
     + Long form:
-        Get(<ProductType>, <SubjectType>, subject)
+        Get(<OutputType>, <InputType>, input)
 
     + Short form
-        Get(<ProductType>, <SubjectType>(<constructor args for subject>))
+        Get(<OutputType>, <InputType>(<constructor args for input>))
 
     The long form supports providing type information to the rule engine that it could not otherwise
-    infer from the subject variable [1]. Likewise, the short form must use inline construction of the
-    subject in order to convey the subject type to the engine.
+    infer from the input variable [1]. Likewise, the short form must use inline construction of the
+    input in order to convey the input type to the engine.
```

Comment:
To me, this part seems immediately more readable, and demonstrates that it is a good paradigm to adopt.